### PR TITLE
[WIP] #15990 - HTTP client auto redirects

### PR DESCRIPTION
**Pull request by RomanIakovlev**
This PR adds support for Akka HTTP client to automatically follow the redirects, if it's allowed by RFC. In its current form, it supports redirects to the same host:port, but not the cross-origin redirects, because it re-uses the same underlying `OutgoingConnectionBlueprint` instance as the one being requested by the HTTP client API user originally.

It's not yet fully tested because of the aforementioned limitation (in other words, I don't know if it makes sense to proceed with the current approach), however some tests are available in the `akka.http.impl.engine.client.RedirectTestSpec`. Most basic cases are covered. Particularly, the last one (marked as ignored) highlights the problem with cross-origin redirects.

The settings API (scaladsl/javadsl) and logic related to change of HTTP methods based on the redirect's status code might need some more work.

@ktoso please review.

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @RomanIakovlev,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -282,14 +307,14 @@ akka.http {
...
-    max-uri-length             = 2k
-    max-method-length          = 16
+    max-uri-length = 2k
+    max-method-length = 16
     max-response-reason-length = 64
-    max-header-name-length     = 64
-    max-header-value-length    = 8k
-    max-header-count           = 64
-    max-chunk-ext-length       = 256
-    max-chunk-size             = 1m
```
Comment (ktoso) [inline akka-http-core/src/main/resources/reference.conf]: Please don't :)

Comment (RomanIakovlev) [inline akka-http-core/src/main/resources/reference.conf]: Oops, using auto-format on this file was not the best idea... :) Fixed.

File: akka-http-core/src/main/resources/reference.conf [hunk 2]
```diff
@@ -170,6 +170,31 @@ akka.http {
     parsing {
       # no overrides by default, see `akka.http.parsing` for default values
     }
+
+    # Automatic redirect settings
+    redirect {
+      # `same-origin` and `cross-origin` contain the same set of settings for
```
Comment (ktoso) [inline akka-http-core/src/main/resources/reference.conf]: excellent, thanks - separating config like this looks very good.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    else false
+  }
+
+  trait OpOutPrinter[F, _ <: Outlet[F]] {
+    def print(t: F): Unit
+  }
+
+  trait OpInPrinter[T <: Inlet[_]] {
+    def print(): Unit
+  }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: I'll see later if we really need those but in general let's not add ad-hoc in-middle-of-other-class type classes for printing :)

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: (Can stay here while we're discussing the PR still, no worries)

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+class RedirectSupportStage(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) extends GraphStage[BidiShape[HttpRequest, HttpRequest, HttpResponse, HttpResponse]] {
+  /**
+    * in1 is where the original messages come from
+    */
+  val requestIn: Inlet[HttpRequest] = Inlet("Input1")
+
+  /**
+    * in2 is the responses to original messages. If it's a redirect, its not forwarded upstream, but goes downstream
+    */
+  val responseIn: Inlet[HttpResponse] = Inlet("Input2")
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: Please use this pattern for port names: `RedirectSupportStage.responseIn` / `RedirectSupportStage.requestOut` etc

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: (If something fails, the name is only printed - so it's more helpful to include names like this, we're trying to standardise on this pattern now)

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+      }
+
+      def doPull[T](inlet: Inlet[T])(implicit ev: OpInPrinter[Inlet[T]]) = {
+        if (!isClosed(inlet)) {
+          pull(inlet)
+        }
+        ev.print()
+      }
+
+      var inFlight = Vector.empty[HttpRequest]
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: It doesn't seem like we're limiting the max nr of in flight messages; would that be a good idea to add here?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        override def onPush(): Unit = {
+          printy("responseIn: onPush")
+          val response = grab(responseIn)
+          val maybeRedirect = redirectMapper(settings, inFlight.head, response)
+          if (maybeRedirect.isDefined) {
+            val redirect = maybeRedirect.get
+            printy(loopDetector.toString())
+            val loops = loopDetector(inFlight.head)
+            if (loops.contains(redirect.uri)) {
+              throw InfiniteRedirectLoopException
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: I forgot which is better - throw or fail the stage? 
The answer depends on if we want to log this failure... I guess we do.
// cc @drewhk

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+import akka.http.scaladsl.settings.ClientAutoRedirectSettings.HeadersForwardMode.{All, Only, Zero}
+import akka.stream._
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+
+import scala.language.higherKinds
+import scala.util.control.NoStackTrace
+
+object RedirectSupportStage {
+  type RedirectMapper = (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest]
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: Let's make it a normal class instead perhaps?
Note that you could: 

```
case class RedirectMapper(a: ClientAutoRedirectSettings, b: HttpRequest, c: HttpResponse) 
extends (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest] {
 def apply...
```

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        /**
+          * Called when the input port has a new element available. The actual element can be retrieved via the
+          * [[GraphStageLogic.grab()]] method.
+          */
+        override def onPush(): Unit = {
+          printy("responseIn: onPush")
+          val response = grab(responseIn)
+          val maybeRedirect = redirectMapper(settings, inFlight.head, response)
+          if (maybeRedirect.isDefined) {
+            val redirect = maybeRedirect.get
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: instead of `isDefined + get` please use `maybeRedirect match { case Some(...`

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala (hunk 6 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: hm, also - would it be useful to include the url in the exception message?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+          * be called for this port.
+          */
+        override def onDownstreamFinish(): Unit = {
+          printy("responseOut: onDownstreamFinish")
+          cancel(responseIn)
+        }
+      })
+
+    }
+}
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: note to self: re-read again

File: akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+object ClientAutoRedirectSettingsImpl extends SettingsCompanion[ClientAutoRedirectSettingsImpl]("akka.http.client.redirect") {
+
+  override def fromSubConfig(root: Config, inner: Config): ClientAutoRedirectSettingsImpl = {
+    val sameOriginConfig = inner.withFallback(root.getConfig(prefix)).getConfig("same-origin")
+    val crossOriginConfig = inner.withFallback(root.getConfig(prefix)).getConfig("cross-origin")
+    ClientAutoRedirectSettingsImpl(
+      ClientAutoRedirectSettingsItemImpl.fromSubConfig(root, sameOriginConfig),
+      ClientAutoRedirectSettingsItemImpl.fromSubConfig(root, crossOriginConfig))
+  }
+}
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala]: great, thanks

File: akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import akka.http.impl.util.SettingsCompanion
+import akka.http.javadsl.settings.ClientAutoRedirectSettingsItem
+import com.typesafe.config.Config
+
+/** INTERNAL API */
+private[akka] final case class ClientAutoRedirectSettingsImpl(
+  sameOrigin: ClientAutoRedirectSettingsItem,
+  crossOrigin: ClientAutoRedirectSettingsItem) extends akka.http.scaladsl.settings.ClientAutoRedirectSettings
+
+object ClientAutoRedirectSettingsImpl extends SettingsCompanion[ClientAutoRedirectSettingsImpl]("akka.http.client.redirect") {
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala]: Hm, companion with Impl in name? Sure

File: akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala [hunk 12]
```diff
@@ -24,7 +24,8 @@ private[akka] final case class ClientConnectionSettingsImpl(
   requestHeaderSizeHint: Int,
   websocketRandomFactory: () ⇒ Random,
   socketOptions: immutable.Seq[SocketOption],
-  parserSettings: ParserSettings)
+  parserSettings: ParserSettings,
+  redirectSettings: akka.http.scaladsl.settings.ClientAutoRedirectSettings)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: AFAICS keeping binary compatibility will force us to write a constructor the same as the previous one with a default value for the new parameter here.

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: (you can check that by running `mimaReportBinaryIssues`

File: akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala [hunk 13]
```diff
@@ -172,6 +172,7 @@ private[http] object JavaMapping {
   implicit object ClientConnectionSettings extends Inherited[js.ClientConnectionSettings, akka.http.scaladsl.settings.ClientConnectionSettings]
   implicit object ConnectionPoolSettings extends Inherited[js.ConnectionPoolSettings, akka.http.scaladsl.settings.ConnectionPoolSettings]
   implicit object ParserSettings extends Inherited[js.ParserSettings, akka.http.scaladsl.settings.ParserSettings]
+  implicit object ClientAutoRedirectSettings extends Inherited[js.ClientAutoRedirectSettings, akka.http.scaladsl.settings.ClientAutoRedirectSettings]
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: :+1:

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 14]
```diff
@@ -258,7 +258,8 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
   def clientLayer(hostHeader: Host,
                   settings: ClientConnectionSettings,
                   log: LoggingAdapter = system.log): ClientLayer =
-    OutgoingConnectionBlueprint(hostHeader, settings, log)
+    RedirectSupportStage(settings.redirectSettings).atop(OutgoingConnectionBlueprint(hostHeader, settings, log))
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: I think we could put that into the blueprint directly

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+import akka.stream.scaladsl._
+import akka.stream.testkit.Utils
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
+import akka.testkit.AkkaSpec
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+class RedirectTestSpec extends AkkaSpec {
+  implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system).withFuzzing(true))
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: fuzzing is enabled by default (see reference.conf in test scope), you can simply `ActorMaterializer()`

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+    "be able to handle infinite redirect loop" in Utils.assertAllStagesStopped {
+      val (_, serverHostName, serverPort) = TestUtils.temporaryServerHostnameAndPort()
+      val cycleLength = 3
+      val binding = Http().bindAndHandleSync(r ⇒ {
+        val c = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse.toInt
+        HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${(c % cycleLength) + 1}")))
+      }, serverHostName, serverPort)
+
+      val x = Source.fromIterator(() ⇒ Iterator.from(1))
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: hm, wouldn't `single()` be enough? The redirects are happening internally, right?

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+        .take(cycleLength)
+        .map(id ⇒ HttpRequest(uri = s"/r$id"))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
+        .runWith(Sink.head)
+
+      a[RedirectSupportStage.InfiniteRedirectLoopException.type] should be thrownBy Await.result(x, 3.second)
+
+      binding.futureValue.unbind()
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: `futureValue` on the unbind as well :)

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+          p.headers.find(_.is("etag")).map(_.value.replace("\"", "").toInt).getOrElse(0)
+        })
+        .runFold(0)(_ + _)
+
+      result.futureValue(PatienceConfig(1000.seconds)) shouldEqual (N + 1) * (N / 2)
+
+      binding.futureValue.unbind()
+    }
+
+    "be able to forward selected headers to redirected requests in different origin" /*in Utils.assertAllStagesStopped */ ignore {
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: I'll check out the branch and investigate this one then; likely not this week though – easter happened.
Sorry for the delay!

**Comment by RomanIakovlev** [comment]
Thanks a bunch for the review! I'll have a break for a couple of days now, and will address your findings afterwards.

**Comment by ktoso** [comment]
Excellent, thank you! Your work looks very through so far, looking forward to figuring out the missing bits and pushing it over the finish line :) Have a nice easter, cheers :rabbit:

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala (hunk 14 above)
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: I can think of some more aspects that'll come up (like caching or proxying). Should they all be baked into the main blueprint?

Comment (RomanIakovlev) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: I thought about those other aspects as well. Maybe it makes sense to introduce something like `HighLevelHttpClient`, to wrap all those aspects and conditionally enable or disable them separately based on the settings? Of course, this potential `HighLevelHttpClient` is bigger than this PR, and there's already an issue for it: https://github.com/akka/akka/issues/16856. I just think this particular place, i.e. `atop(OutgoingConnectionBlueprint)`, is the most appropriate place to add such high-level client.

Every such feature incur some overhead, and client doesn't have to pay the price for something he/she doesn't need, like these redirects, for example. In this particular case, I think I can check the settings on both `cross-origin` and `same-origin` branches, and if both are disabled, don't add this `RedirectSupportStage` on top of the `OutgoingConnectionBlueprint` at all.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+              None
+            }
+          })
+        }
+      } else {
+        None
+      }
+    }
+
+  def sameOrigin(u1: Uri, u2: Uri): Boolean = {
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]:
```scala
def sameOrigin(u1: Uri, u2: Uri) =
  u2.isRelative || u1.isAbsolute && u1.authority == u2.authority
```

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+      case MovedPermanently | Found | TemporaryRedirect | PermanentRedirect if m == GET || m == HEAD => Some(m)
+      case SeeOther => Some(GET)
+      case _ => None
+    }
+  }
+
+  val defaultRedirectMapper: RedirectMapper =
+    (s, req, resp) => {
+      if (resp.status.isRedirection) {
+        inferMethod(req, resp).flatMap { method =>
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: The nested `flatMap`s could be replaced by a for-loop (untested):

```scala
def isAllowed(location: Uri) = {
  val setting =
    if (resp.uri.isRelative || req.uri.isAbsolute && req.uri.authority == resp.uri.authority) s.sameOrigin
    else s.crossOrigin
  setting.getAllow
}

for {
  method <- inferMethod(req, resp)
  location <- resp.header[Location]
  if isAllowed(location.uri)
} yield {
  val headers = item.getForwardHeaders match {
    case All => req.headers
    case Zero => Nil
    case Only(hs) => req.headers.filter(h => hs.contains(h.name))
  }
  req.copy(uri = location.uri, headers = headers, method = method)
}
```

File: akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala (hunk 12 above)
Comment (RomanIakovlev) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: Speaking of binary compatibility, I'm currently struggling with another class, namely `akka.http.impl.settings.ClientConnectionSettingsImpl`. I've been able to create a constructor without the newly added parameter, however `mimaReportBinaryIssues` also complains about other case classes methods like `copy` and `apply`. AFAIK those are compiler generated and I can't override them. Or can I? And should I?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: The settings are very smartly structured to allow us to use case classes while retaining public API binary compatibility, you'll notice that:

```scala
/** INTERNAL API */
private[akka] final case class ClientConnectionSettingsImpl(
```

So noone is supposed to touch this class => we can change it and filter out MiMa warnings that it has about it.

Comment (RomanIakovlev) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: Thanks, it's good to know! Then how to mark these issues as ignored? 

Besides those I only have 2 errors in binary compatibility:
```
[error]  * abstract method getRedirectSettings()akka.http.javadsl.settings.ClientAut
oRedirectSettings in class akka.http.javadsl.settings.ClientConnectionSettings is pr
esent only in current version
```
and
```
[error]  * abstract method redirectSettings()akka.http.scaladsl.settings.ClientAutoR
edirectSettings in class akka.http.scaladsl.settings.ClientConnectionSettings is pre
sent only in current version
```
These are the newly added methods. Shouldn't they also be ignored for binary compatibility purposes? I mean, adding a method shouldn't impact the binary compatibility, should it?

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala]: > I mean, adding a method shouldn't impact the binary compatibility, should it?

It depends. Imagine you've added an abstract method (like we do here), to a class which users are meant to extend. Suddenly you broke their code when they try to upgrade.

In this case the classes are lined up such that we can indeed silence this error – as only we implement this interface.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala (hunk 17 above)
Comment (RomanIakovlev) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: Do you mean `binding.futureValue.unbind().futureValue`? I don't see this in previously existing specs using the same approach.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala (hunk 5 above)
Comment (RomanIakovlev) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: This is a good question. I'm still not sure if this collection `inFlight` is needed at all with current approach. I've already raised this question [before](https://github.com/akka/akka/issues/15990#issuecomment-198074583). I think, current implementation of the push-pull logic prevents more than 1 request in flight, i.e. I don't pull from `requestIn` before I send the reply to the previous request to the `responseOut`. It is also consistent with the in flight numbers I see during tests.

So with this current approach (always one request in flight), the `inFlight` collection is not needed. I've implemented it like this because I was a bit confused on this topic, but now it starts to become clearer. I think I should either just have 1 item in flight, and keep it in an `Option` instead of `Vector`, or change push-pull logic and truly support multiple, if limited, number of requests in flight. Which approach is preferable?

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala (hunk 17 above)
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: Then those other places are likely wrong :)

**Comment by ktoso** [comment]
Was not able to get to this yet, slipping to 2.4.4 (which will be released whenever we have javadsl ready and this likely too)

**Comment by ktoso** [comment]
Refs #15990

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+      }
+    }
+
+    def sameOrigin(u1: Uri, u2: Uri): Boolean = {
+      u2.isRelative || u1.isAbsolute && u1.authority == u2.authority
+    }
+
+    def apply(s: ClientAutoRedirectSettings, req: HttpRequest, resp: HttpResponse): Option[HttpRequest] = {
+      for {
+        location <- resp.header[Location] if resp.status.isRedirection
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: I'd keep the cheap `if` check outside the for-loop.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+        method <- inferMethod(req, resp)
+        item = if (sameOrigin(req.uri, Uri(location.value))) s.sameOrigin else s.crossOrigin
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: `Uri(location.value)` can be replaced with `location.uri`.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 23]
```diff
@@ -0,0 +1,299 @@
...
+  implicit val ResponseInPrinterResp = new OpInPrinter[Inlet[HttpResponse]] {
+    override def print(): Unit = {
+      printy("\tpulling from responseIn")
+    }
+  }
+
+  def apply(settings: ClientAutoRedirectSettings) =
+    BidiFlow.fromGraph(new RedirectSupportStage(settings, DefaultRedirectMapper()))
+
+  def apply(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) =
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: Use default value for `redirectMapper` and remove the above method?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+          * Called when the input port has a new element available. The actual element can be retrieved via the
+          * [[GraphStageLogic.grab()]] method.
+          */
+        override def onPush(): Unit = {
+          printy("responseIn: onPush")
+          val response = grab(responseIn)
+          redirectMapper(settings, inFlight.head, response) match {
+            case Some(redirect) =>
+              printy(loopDetector.toString())
+              val loops = loopDetector(inFlight.head)
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: There should be no loops in here. ;) Rename to `followed`?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+          case All => req.headers
+          case Zero => Nil
+          case Only(hs) => req.headers.filter(h => hs.exists(h.is))
+        }
+        req.withUri(location.value).withHeaders(headers).withMethod(method)
+      }
+    }
+  }
+
+  case class InfiniteRedirectLoopException(loop: immutable.Seq[Uri])
```
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: Just `RedirectLoopException`?

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+        } else {
+          HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${c + 1}")))
+        }
+      }, serverHostName, serverPort)
+
+      val N = 100
+      val result = Source.fromIterator(() ⇒ Iterator.from(1))
+        .take(N)
+        .map(id ⇒ HttpRequest(uri = s"/r$id"))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
```
Comment (2beaucoup) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: Would be nice to use `RedirectSupportStage` directly ISO `Http().outgoingConnection` in these tests so we can get rid of the bindings.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala (hunk 21 above)
Comment (RomanIakovlev) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: @2beaucoup In case the `if` is outside the `for` comprehension, I'd have to step out from the `Option` monad and provide the `else` block. I don't think it looks better. Here, have a look:

      if (resp.status.isRedirection) {
        for {
          location <- resp.header[Location]
          method <- inferMethod(req, resp)
          item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
          if item.allow
        } yield {
          val headers = item.forwardHeaders match {
            case All => req.headers
            case Zero => Nil
            case Only(hs) => req.headers.filter(h => hs.exists(h.is))
          }
          req.withUri(location.value).withHeaders(headers).withMethod(method)
        }
      }
      else {
        None
      }
    }

Instead of:

      for {
        location <- resp.header[Location] if resp.status.isRedirection
        method <- inferMethod(req, resp)
        item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
        if item.allow
      } yield {
        val headers = item.forwardHeaders match {
          case All => req.headers
          case Zero => Nil
          case Only(hs) => req.headers.filter(h => hs.exists(h.is))
        }
        req.withUri(location.value).withHeaders(headers).withMethod(method)
      }

I'd rather keep the existing version in this case. But as you wish. Let me know if you still prefer to have the `if` outside of the `for`.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala (hunk 26 above)
Comment (RomanIakovlev) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: @2beaucoup Sorry, I don't understand your proposal. Could you please clarify?

**Comment by RomanIakovlev** [comment]
One thing has just occurred to me. Please correct me if I'm wrong anywhere in my reasoning.

Every `OutgoingConnectionBlueprint` instance (further referred to just as "outgoing connection") has maximum 1 corresponding TCP connection at any time. This means, one outgoing connection can only be connected to 1 host. This, in turn, means it's impossible to implement cross-origin redirects inside single outgoing connection.

Next, it seems that all of the existing Akka HTTP client features are implemented as graphs. To send request to another origin means to instantiate the outgoing connection inside another outgoing connection. Or, in other words, to support cross-origin redirects we must change the graph shape in the runtime. Which is not possible, at least currently, if I understand Akka Streams correctly.

To summarize, I think such high-level HTTP client features as redirects support cannot be implemented purely in Akka Streams terms. I think it's should possible to achieve the desired behavior using pure actors, but I'm not sure it's a good way to go forward. Or, maybe it's possible to come up with some creative solution, like changing the graph shape at runtime somehow. 

I hope core team members could shed some light here.

**Comment by 2beaucoup** [comment]
You are right @RomanIakovlev and that's one more reason to keep this kind of functionality out of the low level blueprints (and even the default connection methods in `Http`?).

I think your stage should work for cross origin redirects by using it atop `Http.superPool`. Atop `outgoingConnection` and `cachedHostConnectionPool` it should work at least for same origin redirects.

File: akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala (hunk 26 above)
Comment (2beaucoup) [inline akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala]: Sure. `outgoingConnection` produces a `Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]]`. If you provide such a flow yourself you don't need to to bind/unbind:
```scala
val flow = RedirectSupportStage() join Flow[HttpRequest].map(r ⇒ {
  val c = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse.toInt
  if (c % 2 == 0) {
    HttpResponse(entity = c.toString)
  } else {
    HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${c + 1}")))
  }
})
val N = 100
val result = Source.fromIterator(() ⇒ Iterator.from(1))
  .take(N)
  .map(id ⇒ HttpRequest(uri = s"/r$id"))
  .via(flow)
  .mapAsync(4)(_.entity.toStrict(1.second))
  .map { r ⇒ val s = r.data.utf8String; log.debug(s); s.toInt }
  .runFold(0)(_ + _)
```

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala (hunk 21 above)
Comment (2beaucoup) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala]: I agree that the pure monadic style looks better. But the suggested change saves the unnecessary location header lookup which will always return `None` for all non-redirect responses and is more expensive to calculate than `isRedirection `.

If only if-statements could come first in for-loops. ;)

**Comment by drewhk** [comment]
I think on the per-request API level it can be implemented on the top using Future composition by simply recovering from the case where a redirect has been found with another request Future using flatMap.

In the case of Flow based APIs, a loop might be necessary, reinjecting redirected requests to the wrapped flow (with an upper limit) using the graph DSL

**Comment by ktoso** [comment]
Hey there, thanks for the nice chat on ScalaDays :-)
AFAIR we agreed to close it for now, but I'd hope you'd be able to give it another shot given what we've learned from the initial attempt?

A contrib of this feature would be very awesome!
